oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java

```diff
@@ -0,0 +1,186 @@
+            Supplier<String> authorizationSupplier) {
+        clientAuthorization = ClientAuthorization.ofBasicAuthorization(authorizationSupplier);
+        return (T) this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     * @param authorizationType One of the registered HTTP authentication schemes as per
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    @SuppressWarnings("unchecked")
+    public final T clientCredentials(
+            Supplier<? extends Map.Entry<String, String>> credentialsSupplier, String authorizationType) {
+        clientAuthorization = ClientAuthorization.ofCredentials(credentialsSupplier, authorizationType);
+        return (T) this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     */
+    @SuppressWarnings("unchecked")
+    public final T clientCredentials(
+            Supplier<? extends Map.Entry<String, String>> credentialsSupplier) {
+        clientAuthorization = ClientAuthorization.ofCredentials(credentialsSupplier);
+        return (T) this;
+    }
+
+    /**
+     * A period when the token should be refreshed proactively prior to its expiry.
+     */
+    @SuppressWarnings("unchecked")
+    public final T refreshBefore(Duration refreshBefore) {
+        this.refreshBefore = requireNonNull(refreshBefore, "refreshBefore");
+        return (T) this;
+    }
+
+    public final Duration refreshBefore() {
+        return refreshBefore;
+    }
+
+    @Nullable
+    public final Supplier<AccessTokenCapsule> tokenSupplier() {
+        return tokenSupplier;
+    }
+
+    /**
+     * A {@link Supplier} to load Access Token from, to be able to restore the previous session. OPTIONAL.
+     */
+    @SuppressWarnings("unchecked")
+    public final T tokenSupplier(
+            Supplier<AccessTokenCapsule> tokenSupplier) {
+        this.tokenSupplier = requireNonNull(tokenSupplier, "tokenSupplier");
+        return (T) this;
+    }
+
+    @Nullable
+    public final Consumer<AccessTokenCapsule> tokenConsumer() {
+        return tokenConsumer;
+    }
+
+    /**
+     * A {@link Consumer} to store Access Token to, to be able restore the previous session. OPTIONAL.
+     */
+    @SuppressWarnings("unchecked")
+    public final T tokenConsumer(
+            Consumer<AccessTokenCapsule> tokenConsumer) {
```

Comment:
- Could merge two lines.
- Global comment `Consumer<? super AccessTokenCapsule>` and `Supplier<? extends AccessTokenCapsule>`

Reply:
Fixed